### PR TITLE
Fix compile constraints by compile per operating system and using `uv` 

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues
Workaround for the problem described here: https://github.com/napari/napari/pull/6858#issuecomment-2073584997

# Description

Use `uv` instead of pip-tools for faster compilation and support of targeting python version and os.

### CODE REVIEW

**Review by brisvag — COMMENTED** [review]
So this results in  12 constrain files where we had 4 before, right?

**Comment by Czaki** [comment]
For current state 16 😜 
But the i would like to check if they really different. And maybe it will only 8 or even less

**Comment by Czaki** [comment]
pushing this forward requires to fix handling macos-arm in uv https://github.com/astral-sh/uv/pull/3267

File: pyproject.toml [hunk 1]
```diff
@@ -91,7 +91,7 @@ Documentation = "https://napari.org"
 
 [project.optional-dependencies]
 pyside2 = [
-    "PySide2>=5.13.2,!=5.15.0 ; python_version < '3.12'",
+    "PySide2>=5.13.2,!=5.15.0 ; python_version < '3.12' and platform_machine == 'x86_64'",
```
Comment (psobolewskiPhD) [inline pyproject.toml]:
```suggestion
    "PySide2>=5.13.2,!=5.15.0 ; python_version < '3.11' and platform_machine == 'x86_64'",
```
None of the wheels on pypi support 3.11

**Comment by brisvag** [comment]
> pushing this forward requires to fix handling macos-arm in uv [astral-sh/uv#3267](https://github.com/astral-sh/uv/pull/3267)

released!

**Comment by Czaki** [comment]
@brisvag it is ready for review

File: .github/workflows/reusable_run_tox_test.yml [hunk 2]
```diff
@@ -49,7 +49,7 @@ jobs:
       PYVISTA_OFF_SCREEN: True
       MIN_REQ: ${{ inputs.min_req }}
       FORCE_COLOR: 1
-      PIP_CONSTRAINT: resources/constraints/constraints_py${{ inputs.python_version }}${{ inputs.min_req && '_min_req' }}${{ inputs.constraints_suffix }}.txt
+      PIP_CONSTRAINT: resources/constraints/constraints_py${{ inputs.python_version }}${{ (((inputs.platform == 'macos-latest') && '_macos_arm') || '') }}${{ inputs.min_req && '_min_req' }}${{ inputs.constraints_suffix }}.txt
```
Comment (brisvag) [inline .github/workflows/reusable_run_tox_test.yml:52]: please add a comment on this, it's really hard to parse xD

File: .github/workflows/upgrade_test_constraints.yml [hunk 3]
```diff
@@ -145,35 +145,22 @@ jobs:
...
-          flags=""
+          set -x
+          pip install -U uv
+          flags="--quiet"
           # Explanation of below commands
-          # python3.8 -m piptools compile - call pip-compile but ensure proper interpreter
+          # uv pip compile --python-version 3.9 - call uv pip compile but ensure proper interpreter
           # --upgrade upgrade to the latest possible version. Without this pip-compile will take a look to output files and reuse versions (so will ad something on when adding dependency.
-          # -o resources/constraints/constraints_py3.8.txt - output file
+          # -o resources/constraints/constraints_py3.9.txt - output file
```
Comment (brisvag) [inline .github/workflows/upgrade_test_constraints.yml:163]: this update to 3.9 is a leftover from when we dropped 3.8 in #6826?

File: .github/workflows/upgrade_test_constraints.yml [hunk 4]
```diff
@@ -189,24 +176,16 @@ jobs:
...
-          flags+=" --resolver=backtracking"
-
           for pyv in 3.9 3.10 3.11 3.12; do
-            python${pyv}  -m pip install -U pip pip-tools
-            python${pyv}  -m piptools compile --upgrade -o $constraints/constraints_py${pyv}.txt  $pyproject_toml $constraints/version_denylist.txt ${flags}
-            python${pyv}  -m piptools compile --upgrade -o $constraints/constraints_py${pyv}_pydantic_1.txt  $pyproject_toml $constraints/version_denylist.txt $constraints/pydantic_le_2.txt ${flags}
+            uv pip compile --python-version ${pyv} --upgrade --output-file $constraints/constraints_py${pyv}.txt  $pyproject_toml $constraints/version_denylist.txt ${flags}
+            uv pip compile --python-version ${pyv} --upgrade --output-file $constraints/constraints_py${pyv}_pydantic_1.txt  $pyproject_toml $constraints/version_denylist.txt $constraints/pydantic_le_2.txt ${flags}
+            uv pip compile --python-platform aarch64-apple-darwin  --python-version ${pyv} --upgrade --output-file $constraints/constraints_py${pyv}_macos_arm.txt  $pyproject_toml $constraints/version_denylist.txt ${flags}
```
Comment (brisvag) [inline .github/workflows/upgrade_test_constraints.yml:182]: So we no longer have a whole matrix, just a special case for macos?

File: .github/workflows/upgrade_test_constraints.yml (hunk 3 above)
Comment (Czaki) [reply .github/workflows/upgrade_test_constraints.yml:163]: yes

File: .github/workflows/upgrade_test_constraints.yml (hunk 4 above)
Comment (Czaki) [reply .github/workflows/upgrade_test_constraints.yml:182]: Yes. I started testing of post compile merge of files into one, but it will prevent from implementing the feature of updating constraints only for a given package. So I ended with an additional macOS arm file. I hope that we could drop it shortly.

File: .github/workflows/reusable_run_tox_test.yml (hunk 2 above)
Comment (Czaki) [reply .github/workflows/reusable_run_tox_test.yml:52]: could you check now?

**Comment by psobolewskiPhD** [comment]
Pretty wild that the action used to take ~9 min and now it's 25 s.

**Comment by Czaki** [comment]
> Pretty wild that the action used to take ~9 min and now it's 25 s.

This speedup is a reason why I started these migration.

File: .github/workflows/reusable_run_tox_test.yml (hunk 2 above)
Comment (psobolewskiPhD) [reply .github/workflows/reusable_run_tox_test.yml:52]: I like the comment!

**Review by psobolewskiPhD — APPROVED** [review]
Seems good to me.
Big savings of CI time, so I'd say lets go with it!
🚀

**Comment by brisvag** [comment]
Great work!
